code/webauthn/frontend/utils/webauthn.ts

```diff
@@ -0,0 +1,117 @@
+import { startAuthentication } from '@simplewebauthn/browser';
+import { bufferToHex, parseHex, arrayBufferToHex, bufferFromBase64url } from './string';
+import { defaultAbiCoder } from 'ethers/lib/utils';
+import { EIP712Signer, type Provider, utils } from 'zksync-ethers';
+import type { AuthenticatorAssertionResponseJSON } from '@simplewebauthn/types';
+import { BigNumber } from 'ethers';
+import type { TransactionRequest } from 'zksync-ethers/src/types';
+import * as cbor from 'cbor';
+
+export async function authenticate(challenge: string) {
+  const resp = await fetch('http://localhost:3000/api/get-authentication-options', {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+      Accept: 'application/json',
+    },
+    body: JSON.stringify({ challenge }),
+  });
+  const data = await resp.json();
+  const options = data.options;
+  const authResp = await startAuthentication(options);
+  return authResp.response;
+}
+
+export async function signAndSend(
+  provider: Provider,
+  transaction: TransactionRequest,
+  authenticationResponse: AuthenticatorAssertionResponseJSON
+) {
+  const signature = await getSignatureFromAuthResponse(authenticationResponse);
+  const newTx = {
+    ...transaction,
+    customData: {
+      ...transaction.customData,
+      customSignature: signature,
+    },
+  };
+  return await send(newTx, provider);
+}
+
+export async function send(transaction: TransactionRequest, provider: Provider) {
+  const sentTx = await provider.sendTransaction(utils.serialize(transaction));
+  const txReceipt = await sentTx.wait();
+  return txReceipt;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export async function getSignatureFromAuthResponse(response: any) {
+  const authenticatorDataBuffer = bufferFromBase64url(response.authenticatorData);
+  const clientDataBuffer = bufferFromBase64url(response.clientDataJSON);
+  const rs = getRS(response.signature);
+  const signature = defaultAbiCoder.encode(
+    ['bytes', 'bytes', 'bytes32[2]'],
+    [authenticatorDataBuffer, clientDataBuffer, rs]
+  );
+  return signature;
+}
+
+export function getDataToSign(transaction: TransactionRequest) {
+  const signedTxHash = EIP712Signer.getSignedDigest(transaction);
+  return signedTxHash;
+}
+
+export function getRS(signatureBase64Url: string): Array<BigNumber> {
+  const n = BigNumber.from('0xFFFFFFFF00000000FFFFFFFFFFFFFFFFBCE6FAADA7179E84F3B9CAC2FC632551');
+  const halfN = n.div(2);
+  const signatureBuffer = bufferFromBase64url(signatureBase64Url);
+  const signatureParsed = derToRS(signatureBuffer);
+  const firstHex = arrayBufferToHex(signatureParsed[0]);
+  const first = BigNumber.from('0x' + firstHex);
+  const second = BigNumber.from('0x' + arrayBufferToHex(signatureParsed[1]));
+  const sig: Array<BigNumber> = [first, second];
+
+  if (sig[1].gt(halfN)) {
+    sig[1] = n.sub(sig[1]);
```

Comment:
Aah, so this is the S normalization. Might be good to explain a bit of what's going on here as I had to trace a bit to find this!